lib/auth/auth_with_roles.go

```diff
@@ -1329,6 +1339,208 @@ func (a *ServerWithRoles) ListResources(ctx context.Context, req proto.ListResou
+			case match:
+				resp.Resources = append(resp.Resources, r)
+			}
+			// repectLimit is true only if we do not require the fake pagination field.
+			if len(resp.Resources) == limit && respectLimit {
+				resp.NextKey = continueKey
+				return len(resp.Resources), ErrDone
+			}
+			return len(resp.Resources), nil
+		},
+	)
+	return resp, trace.Wrap(err)
+}
+
+// iterateKubernetesPods creates a new Kubernetes Client with temporary user
+// certificates and iterates through the returned Kubernetes Pods.
+// For each Pod discovered, the fn function is called to decide the action.
+// Kubernetes continue key is a base64 encoded json payload with the resource
+// version of the request. In order to resume the operation when using the paginated
+// mode, Teleport respects the Kubernetes Continue Key and will return it to the client
+// as a NextKey.
+// In order to have the expected behavior Teleport must respect the ContinueKey and
+// cannot manipulate it. It means that Teleport needs to manipulate the number of
+// requested items from the Kubernetes Cluster in order to have the expected behavior.
+func (a *ServerWithRoles) iterateKubernetesPods(
+	ctx context.Context,
+	req proto.ListResourcesRequest,
+	respectLimit bool,
+	fn func(types.ResourceWithLabels, string) (int, error),
+) error {
+	c, err := a.getKubernetesClient(ctx, req.KubernetesPodsFilter.Cluster)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	continueKey := req.StartKey
+	itemsAppended := 0
+	for {
+		podList, err := c.CoreV1().Pods(req.KubernetesPodsFilter.Namespace).List(ctx, v1.ListOptions{
+			Limit:    decideLimit(int64(req.Limit), int64(itemsAppended), respectLimit),
+			Continue: continueKey,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		for _, pod := range podList.Items {
+			resource, err := types.NewKubernetesPodV1(
+				types.Metadata{
+					Name:   pod.Name,
+					Labels: pod.Labels,
+				},
+				types.KubernetesPodSpecV1{
+					Namespace: pod.Namespace,
+				},
+			)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			itemsAppended, err = fn(resource, podList.Continue)
+			if errors.Is(err, ErrDone) {
+				return nil
+			} else if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		if len(podList.Continue) == 0 {
+			return nil
+		}
+		continueKey = podList.Continue
+	}
+}
+
+// decideLimit returns the number of items we should request for. If respectLimit
+// is true, it returns the difference between the max number of items and the
+// number of items already included in the response.
+// If false, returns the max number of items.
+func decideLimit(limit, items int64, respectLimit bool) int64 {
+	if respectLimit {
+		return limit - items
+	}
+	return limit
+}
+
+// getKubernetesClient creates a new Kubernetes client configured with temporary
+// certificates bounded with `search_as_role` roles.
+func (a *ServerWithRoles) getKubernetesClient(ctx context.Context, kubeCluster string) (*kubernetes.Clientset, error) {
```

Comment:
Can we split the responsibilities of `ServerWithRoles` and move the `getKubernetesClient` kube related functions to other object inside kube package ?